cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,205 @@
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteUser(message string) {
+	l.write(l.userPrefix, message)
 }
 
-// LoggerInterface is logger abstraction
-type LoggerInterface interface {
-	Log(message string)
+// WriteDebug writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteDebug(message string) {
+	l.write(l.debugPrefix, message)
 }
 
-// NewStdoutLogger creates a new logger which uses prefix for all messages logged.
-// All messages are sent to stdout.
-func NewStdoutLogger(prefix string) LoggerInterface {
-	return stdoutLogger{prefix: prefix}
+// WriteTrace buffers the message for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteTrace(message string) {
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
+
+	l.trace = append(l.trace, message)
 }
 
-type stdoutLogger struct{ prefix string }
+// WriteMetric keeps non-nil fields from metric for inclusion in ReadOutputInfo().
+// Elements of list fields are appended to the underlying view.
+func (l *defaultToolLogger) WriteMetric(metric *pb.OutputInfo) {
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
 
-// Log logs a message
-func (l stdoutLogger) Log(message string) {
-	fmt.Printf("%s %s\n", l.prefix, newLogEntry(message))
+	proto.Merge(l.outputInfo, metric)
 }
 
-// NewDefaultLogger creates a new logger that sends all messages to the default logger.
-func NewDefaultLogger() LoggerInterface {
-	return defaultLogger{}
+// Returns a view comprised of:
+//   - Calls to WriteMetric
+//   - All user, debug, and trace logs. User and debug logs are appended into a single
+//     member of OutputInfo.SerialLogs; each trace log is a separate member.
+// All buffers are cleared when this is called. In other words, a subsequent call to
+// ReadOutputInfo will return an empty object.
+func (l *defaultToolLogger) ReadOutputInfo() *pb.OutputInfo {
+	// Locking since ReadOutputInfo has a side effect of clearing the internal state.
+	l.mutationLock.Lock()
+	defer l.mutationLock.Unlock()
+
+	// Prepend stdout logs (user, debug) to the trace logs.
+	var combinedTrace []string
+	if l.userAndDebugBuffer.Len() > 0 {
+		combinedTrace = []string{l.userAndDebugBuffer.String()}
+		l.userAndDebugBuffer.Reset()
+	}
+	ret := &pb.OutputInfo{SerialOutputs: append(combinedTrace, l.trace...)}
+	l.trace = []string{}
+	proto.Merge(ret, l.outputInfo)
+	l.outputInfo = &pb.OutputInfo{}
+	return ret
 }
 
-type defaultLogger struct{}
+// write a message to the underlying logger, and buffer it for inclusion in OutputInfo.SerialLogs.
+func (l *defaultToolLogger) write(prefix, message string) {
+	var logLineBuilder strings.Builder
+	// If there's a prefix, ensure it ends with a colon.
+	if prefix != "" {
+		logLineBuilder.WriteString(prefix)
+		if !strings.HasSuffix(prefix, ":") {
+			logLineBuilder.WriteByte(':')
+		}
```

Comment:
Consider updating the prefix on construction if `:` is always appended anyway, to avoid having to do it for each line

Reply:
I'd definitely agree if the struct members were immutable, but given that they could change, it seems most safe (to me) to ensure the correct output when the prefix is written.

That said, I'm not wedded to it. If you prefer doing the formatting once in `NewToolLogger`, I can do that!